### PR TITLE
[SPARK-12363] [SQL] Inline Hive parser into spark sql

### PR DESCRIPTION
**Pull request by nongli**
This is a WIP. This inlines the hive sql grammar parser into spark sql in the hive
subproject. This should eventually be moved into the SQL core project once all the
hive dependencies are removed.

This patch does some of that by cleaning up the hive code to remove much of semantic
analysis. The SBT plugin uses 3.5.2 (the latest in 3.5.x) and this updates the grammar
to support that (from 3.4). Changes are very minor.

### CODE REVIEW

**Comment by rxin** [comment]
I think you meant https://issues.apache.org/jira/browse/SPARK-12362, not 12363?

**Comment by rxin** [comment]
Not sure if you have already done it. If not, we should make sure we make it clear in source code that this is coming from Hive.

File: sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+      throws ParseException {
+    LOG.info("Parsing command: " + command);
+
+    HiveLexerX lexer = new HiveLexerX(new ANTLRNoCaseStringStream(command));
+    TokenRewriteStream tokens = new TokenRewriteStream(lexer);
+    if (ctx != null) {
+      if ( setTokenRewriteStream) {
+        ctx.setTokenRewriteStream(tokens);
+      }
+      lexer.setHiveConf(ctx.getConf());
```
Comment (hvanhovell) [inline sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java]: The lexer only checks the ```HiveConf``` to check if it supports quoted ids. We could move this to our own configuration.

File: sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    }
+    SparkSqlParser parser = new SparkSqlParser(tokens);
+    if (ctx != null) {
+      parser.setHiveConf(ctx.getConf());
```
Comment (hvanhovell) [inline sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java]: The parser only checks the ```HiveConf``` to check if we can use SQL11 as identifiers. We could move this to our own configuration.

File: sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+   */
+  public ASTNode parse(String command, Context ctx, boolean setTokenRewriteStream) 
+      throws ParseException {
+    LOG.info("Parsing command: " + command);
+
+    HiveLexerX lexer = new HiveLexerX(new ANTLRNoCaseStringStream(command));
+    TokenRewriteStream tokens = new TokenRewriteStream(lexer);
+    if (ctx != null) {
+      if ( setTokenRewriteStream) {
+        ctx.setTokenRewriteStream(tokens);
```
Comment (hvanhovell) [inline sql/hive/src/main/java/org/apache/spark/sql/parser/ParseDriver.java]: The ```Context.tokenRewriteStream``` is used in HiveQL. If we factor out this and the ```HiveQL``` (see comments below) we can drop the Context entirely.

**Pull request closed without merging**